### PR TITLE
[SPARK-12081] Make unified memory manager work with small heaps

### PR DESCRIPTION
**Pull request by andrewor14**
The existing `spark.memory.fraction` (default 0.75) gives the system 25% of the space to work with. For small heaps, this is not enough: e.g. default 1GB leaves only 250MB system memory. This is especially a problem in local mode, where the driver and executor are crammed in the same JVM. Members of the community have reported driver OOM's in such cases.

**New proposal.** We now reserve 300MB before taking the 75%. For 1GB JVMs, this leaves `(1024 - 300) * 0.75 = 543MB` for execution and storage. This is proposal (1) listed in the [JIRA](https://issues.apache.org/jira/browse/SPARK-12081).

### CODE REVIEW

**Comment by andrewor14** [comment]
@davies @rxin

**Comment by davies** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 1]
```diff
@@ -130,6 +130,8 @@ private[spark] class UnifiedMemoryManager private[memory] (
 
 object UnifiedMemoryManager {
 
+  private val RESERVED_SYSTEM_MEMORY_BYTES = 300 * 1024 * 1024
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala]: you should document what this is and update the classdoc for UnifiedMemoryManager

**Comment by andrewor14** [comment]
Thanks I'm merging into master 1.6

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 2]
```diff
@@ -130,6 +130,12 @@ private[spark] class UnifiedMemoryManager private[memory] (
 
 object UnifiedMemoryManager {
 
+  // Set aside a fixed amount of memory for non-storage, non-execution purposes.
+  // This serves a function similar to `spark.memory.fraction`, but guarantees that we reserve
+  // sufficient memory for the system even for small heaps. E.g. if we have a 1GB JVM, then
+  // the memory used for execution and storage will be (1024 - 300) * 0.75 = 543MB by default.
+  private val RESERVED_SYSTEM_MEMORY_BYTES = 300 * 1024 * 1024
```
Comment (tedyu) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala]: For non-small heap, the above memory doesn't need to be reserved, right ?

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 3]
```diff
@@ -144,8 +150,16 @@ object UnifiedMemoryManager {
...
    */
   private def getMaxMemory(conf: SparkConf): Long = {
-    val systemMaxMemory = conf.getLong("spark.testing.memory", Runtime.getRuntime.maxMemory)
+    val systemMemory = conf.getLong("spark.testing.memory", Runtime.getRuntime.maxMemory)
+    val reservedMemory = conf.getLong("spark.testing.reservedMemory",
+      if (conf.contains("spark.testing")) 0 else RESERVED_SYSTEM_MEMORY_BYTES)
+    val minSystemMemory = reservedMemory * 1.5
+    if (systemMemory < minSystemMemory) {
+      throw new IllegalArgumentException(s"System memory $systemMemory must " +
+        s"be at least $minSystemMemory. Please use a larger heap size.")
```
Comment (jcsalterego) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala]: This comes out as scientific notation, which made me laugh.

**Pull request closed without merging**